tests/test_venv.rs

```diff
@@ -0,0 +1,61 @@
+use std::fs::remove_dir_all;
+use std::process::Command;
+
+#[test]
+fn test_building_example_using_venv() {
+    fn create_venv(name: &str) {
+        Command::new("python")
+            .args(&["-m", "venv", name])
+            .status()
+            .expect("failed to create virtual environment")
+            .success();
+    }
+
+    fn remove_venv(name: &str) {
+        remove_dir_all(name).expect("failed to remove virtual environment");
+    }
+
+    fn is_example_built_using_venv(venv_name: &str, example_name: &str) -> bool {
+        if cfg!(target_os = "windows") {
+            Command::new("cmd")
+                .args(&[
+                    "/C",
+                    format!(
+                        "{0}\\Scripts\\activate.bat && \
+                         cd examples\\{1} && \
+                         python -m pip install --upgrade setuptools-rust && \

```

Comment:
without pre-installing `setuptools-rust` package command
```cmd
python setup.py build
```
fails with
```cmd
running build_ext
running build_rust
error: invalid command 'build_rust'
```
so it should be installed before

Reply:
@lycantropos Actually you should not invoke `setup.py` directly, instead use `pip install` with a `pyproject.toml`. The `word-count` and `rustapi_modules` examples are built with PEP 517, and installing or building them by invoking `setup.py` probably works even worse than it would in a normal project.

Reply:
```cmd
(venv) > python -m pip install --upgrade pip
Requirement already up-to-date: pip in c:\users\azati\clionprojects\pyo3\venv\lib\site-packages (19.1)
(venv) > python -m pip install .
```
gives
```cmd
> python -m pip install .
Processing c:\users\azati\clionprojects\pyo3\examples\word-count
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... done
Building wheels for collected packages: word-count
  Building wheel for word-count (PEP 517) ... error
  ERROR: Complete output from command 'C:\Users\azati\CLionProjects\pyo3\venv\Scripts\python.exe' 'C:\Users\azati\CLionProjects\pyo3\venv\lib\site-packages\pip\_vendor\pep51
7\_in_process.py' build_wheel 'C:\Users\azati\AppData\Local\Temp\tmp1rcvfwh_':
  ERROR: running bdist_wheel
  running build
  running build_py
  running egg_info
  writing word_count.egg-info\PKG-INFO
  writing top-level names to word_count.egg-info\top_level.txt
  writing dependency_links to word_count.egg-info\dependency_links.txt
  reading manifest file 'word_count.egg-info\SOURCES.txt'
  reading manifest template 'MANIFEST.in'
  writing manifest file 'word_count.egg-info\SOURCES.txt'
  running build_ext
  running build_rust
  error: failed to load source for a dependency on `pyo3`

  Caused by:
    Unable to update C:\Users\azati\AppData\Local

  Caused by:
    failed to read `C:\Users\azati\AppData\Local\Cargo.toml`

  Caused by:
    The system cannot find the file specified. (os error 2)
  Traceback (most recent call last):
    File "C:\Users\azati\CLionProjects\pyo3\venv\lib\site-packages\pip\_vendor\pep517\_in_process.py", line 207, in <module>
      main()
    File "C:\Users\azati\CLionProjects\pyo3\venv\lib\site-packages\pip\_vendor\pep517\_in_process.py", line 197, in main
      json_out['return_val'] = hook(**hook_input['kwargs'])
    File "C:\Users\azati\CLionProjects\pyo3\venv\lib\site-packages\pip\_vendor\pep517\_in_process.py", line 141, in build_wheel
      metadata_directory)
    File "C:\Users\azati\AppData\Local\Temp\pip-build-env-n1a5am83\overlay\Lib\site-packages\setuptools\build_meta.py", line 192, in build_wheel
      self.run_setup()
    File "C:\Users\azati\AppData\Local\Temp\pip-build-env-n1a5am83\overlay\Lib\site-packages\setuptools\build_meta.py", line 141, in run_setup
      exec(compile(code, __file__, 'exec'), locals())
    File "setup.py", line 92, in <module>
      'sdist': CargoModifiedSdist,
    File "C:\Users\azati\AppData\Local\Temp\pip-build-env-n1a5am83\overlay\Lib\site-packages\setuptools\__init__.py", line 145, in setup
      return distutils.core.setup(**attrs)
    File "C:\Python35\lib\distutils\core.py", line 148, in setup
      dist.run_commands()
    File "C:\Python35\lib\distutils\dist.py", line 955, in run_commands
      self.run_command(cmd)
    File "C:\Python35\lib\distutils\dist.py", line 974, in run_command
      cmd_obj.run()
    File "C:\Users\azati\AppData\Local\Temp\pip-build-env-n1a5am83\overlay\Lib\site-packages\wheel\bdist_wheel.py", line 192, in run
      self.run_command('build')
    File "C:\Python35\lib\distutils\cmd.py", line 313, in run_command
      self.distribution.run_command(command)
    File "C:\Python35\lib\distutils\dist.py", line 974, in run_command
      cmd_obj.run()
    File "C:\Python35\lib\distutils\command\build.py", line 135, in run
      self.run_command(cmd_name)
    File "C:\Python35\lib\distutils\cmd.py", line 313, in run_command
      self.distribution.run_command(command)
    File "C:\Python35\lib\distutils\dist.py", line 974, in run_command
      cmd_obj.run()
    File "C:\Users\azati\AppData\Local\Temp\pip-build-env-n1a5am83\overlay\Lib\site-packages\setuptools_rust\build_ext.py", line 26, in run
      build_rust.run()
    File "C:\Users\azati\AppData\Local\Temp\pip-build-env-n1a5am83\overlay\Lib\site-packages\setuptools_rust\build.py", line 313, in run
      self.build_extension(ext)
    File "C:\Users\azati\AppData\Local\Temp\pip-build-env-n1a5am83\overlay\Lib\site-packages\setuptools_rust\build.py", line 95, in build_extension
      metadata = json.loads(check_output(metadata_command).decode("utf-8"))
    File "C:\Python35\lib\subprocess.py", line 316, in check_output
      **kwargs).stdout
    File "C:\Python35\lib\subprocess.py", line 398, in run
      output=stdout, stderr=stderr)
  subprocess.CalledProcessError: Command '['cargo', 'metadata', '--manifest-path', 'Cargo.toml', '--format-version', '1']' returned non-zero exit status 101
  ----------------------------------------
  ERROR: Failed building wheel for word-count
  Running setup.py clean for word-count
Failed to build word-count
ERROR: Could not build wheels for word-count which use PEP 517 and cannot be installed directly
```
or it should be made differently?

Reply:
get similar output in WSL
```bash
(venv) $ python -m pip install -U pip
...
Successfully installed pip-19.1
(venv) $ python -m pip install .
Processing /home/ace/pyo3/examples/word-count
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... done
Building wheels for collected packages: word-count
  Building wheel for word-count (PEP 517) ... error
  ERROR: Complete output from command /home/ace/pyo3/examples/word-count/venv/bin/python /home/ace/pyo3/examples/word-count/venv/lib/python3.6/site-packages/pip/_vendor/pep517/_in_process.py build_wheel /tmp/tmp_ald7kvr:
  ERROR: running bdist_wheel
  running build
  running build_py
  running egg_info
  writing word_count.egg-info/PKG-INFO
  writing dependency_links to word_count.egg-info/dependency_links.txt
  writing top-level names to word_count.egg-info/top_level.txt
  reading manifest file 'word_count.egg-info/SOURCES.txt'
  reading manifest template 'MANIFEST.in'
  writing manifest file 'word_count.egg-info/SOURCES.txt'
  running build_ext
  running build_rust
  error: failed to load source for a dependency on `pyo3`

  Caused by:
    Unable to update /

  Caused by:
    failed to read `/Cargo.toml`

  Caused by:
    No such file or directory (os error 2)
  Traceback (most recent call last):
    File "/home/ace/pyo3/examples/word-count/venv/lib/python3.6/site-packages/pip/_vendor/pep517/_in_process.py", line 207, in <module>
      main()
    File "/home/ace/pyo3/examples/word-count/venv/lib/python3.6/site-packages/pip/_vendor/pep517/_in_process.py", line 197, in main
      json_out['return_val'] = hook(**hook_input['kwargs'])
    File "/home/ace/pyo3/examples/word-count/venv/lib/python3.6/site-packages/pip/_vendor/pep517/_in_process.py", line 141, in build_wheel
      metadata_directory)
    File "/tmp/pip-build-env-ir8gxftr/overlay/lib/python3.6/site-packages/setuptools/build_meta.py", line 192, in build_wheel
      self.run_setup()
    File "/tmp/pip-build-env-ir8gxftr/overlay/lib/python3.6/site-packages/setuptools/build_meta.py", line 141, in run_setup
      exec(compile(code, __file__, 'exec'), locals())
    File "setup.py", line 92, in <module>
      'sdist': CargoModifiedSdist,
    File "/tmp/pip-build-env-ir8gxftr/overlay/lib/python3.6/site-packages/setuptools/__init__.py", line 145, in setup
      return distutils.core.setup(**attrs)
    File "/usr/lib/python3.6/distutils/core.py", line 148, in setup
      dist.run_commands()
    File "/usr/lib/python3.6/distutils/dist.py", line 955, in run_commands
      self.run_command(cmd)
    File "/usr/lib/python3.6/distutils/dist.py", line 974, in run_command
      cmd_obj.run()
    File "/tmp/pip-build-env-ir8gxftr/overlay/lib/python3.6/site-packages/wheel/bdist_wheel.py", line 192, in run
      self.run_command('build')
    File "/usr/lib/python3.6/distutils/cmd.py", line 313, in run_command
      self.distribution.run_command(command)
    File "/usr/lib/python3.6/distutils/dist.py", line 974, in run_command
      cmd_obj.run()
    File "/usr/lib/python3.6/distutils/command/build.py", line 135, in run
      self.run_command(cmd_name)
    File "/usr/lib/python3.6/distutils/cmd.py", line 313, in run_command
      self.distribution.run_command(command)
    File "/usr/lib/python3.6/distutils/dist.py", line 974, in run_command
      cmd_obj.run()
    File "/tmp/pip-build-env-ir8gxftr/overlay/lib/python3.6/site-packages/setuptools_rust/build_ext.py", line 26, in run
      build_rust.run()
    File "/tmp/pip-build-env-ir8gxftr/overlay/lib/python3.6/site-packages/setuptools_rust/build.py", line 313, in run
      self.build_extension(ext)
    File "/tmp/pip-build-env-ir8gxftr/overlay/lib/python3.6/site-packages/setuptools_rust/build.py", line 95, in build_extension
      metadata = json.loads(check_output(metadata_command).decode("utf-8"))
    File "/usr/lib/python3.6/subprocess.py", line 336, in check_output
      **kwargs).stdout
    File "/usr/lib/python3.6/subprocess.py", line 418, in run
      output=stdout, stderr=stderr)
  subprocess.CalledProcessError: Command '['cargo', 'metadata', '--manifest-path', 'Cargo.toml', '--format-version', '1']' returned non-zero exit status 101.
  ----------------------------------------
  ERROR: Failed building wheel for word-count
  Running setup.py clean for word-count
Failed to build word-count
ERROR: Could not build wheels for word-count which use PEP 517 and cannot be installed directly
```